### ISSUE TITLE
doc: exact fhmm ? fhmm vs additive fhmm vs additive fhmm for nilm

### ISSUE THREAD

**Issue by yetanotherion** [nilmtk/nilmtk#506]
Hello to all, 
when reading the source code of the exact_fhmm, it seems
it's based on an **additive fhmm** whose hidden states are **decoded** using viterbi's algorithm.

When reading related papers, the exact_fhmm approach is said to have scalability issues: an exact_fhmm for 20 devices with 4 states cannot be computed according to (http://redd.csail.mit.edu/kolter-kddsust11.pdf). These scalability issues can be solved: 
- in general as depicted in http://www.ee.columbia.edu/~sfchang/course/svia-F03/papers/factorial-HMM-97.pdf. In that article, they compare their approach that **scales** vs the **exact** approach (which may be what motivates the **exact** prefix in your code (exact ~ viterbi)).
- more precisely in http://www.zicokolter.com/wp-content/uploads/2015/10/kolter-aistats12.pdf, another inference algorithm is proposed for the nilm use case, which seems to outperform the approach depicted above while still being able to scale.

Due to those scalability issues, the **exact** approach cannot be compared to **these two other approaches** when ~nber_device with more than 2 states~ > some value (the **exact** approach
cannot be computed).

For this reason, the performance than can reach the **exact** approach might be limited, as modelling a 4 state device with 2 only, has a penalty.

Is my understanding correct ?  If so, would you like me to update a related doc accordingly ?

**Comment by oliparson** [comment]
Just to clarify:

The exact FHMM works by creating an equivalent HMM in which each state represents one combination of appliance states. As such, the complexity of exact inference scales exponentially with the number of appliances. For this reason, approximate methods have been proposed which aim to provide good enough solutions given much less computation time. However, NILMTK currently does not provide any approximate algorithms for inference in HMMs.

I don't think it's fair to say that the performance of the exact algorithm is penalised just because you've had to simplify the model to reduce the computation. It's probably clearer to say that the exact approach will always arrive at the right answer, although it might take a very long time to get there.

**Comment by yetanotherion** [comment]
Thanks a lot for this input !
I agree with your rephrasing proposal. 

I raised this issue after reading in multiple papers, that the nilmtk project's FHMM implementation was taken as a benchmark reference for FHMM, without detailing whether the number of states used to model a device was simplified to 2 or not.
For example, in https://arxiv.org/pdf/1507.06594.pdf, the FHMM method for the washing machine had quite a bad result compared to others.

However, if the experiment was run with num_states=2 for the washing machine, such bad results are expected. In that peculiar case, an FHMM disaggregation modelling a washing machine with 4 states might outperform the LSTM as:
"Our LSTM results suggest that LSTMs work best for twostate
appliances but do not perform well on multi-state appliances
such as the dish washer and washing machine."

For this reason, when comparing FHMM disaggregation methods to others, detailing: 
- the number of states used to model devices, and
- the algorithm used to infer the FHMM states,
seem necessary to me.

Adding a section in the documentation regarding that detail, might permit the authors of papers
based on your project, to design benchmarking experiments accordingly.

If you agree on this, I'd be glad to write/help writing such a section.

**Comment by yetanotherion** [comment]
(It actually ends up to whether the following if 
https://github.com/nilmtk/nilmtk/blob/master/nilmtk/disaggregate/fhmm_exact.py#L258
is executed or not in the experiments)

**Comment by oliparson** [comment]
I completely agree that it's good practice to specific both the parameters and inference algorithm used for any model. Indeed, I think the NILMTK FHMM implementation allows you to specify the number of states for each individual appliance. However, I'm not sure whether the NILMTK docs are the right place for this kind of research methodology guide. Maybe a better place might be the newly created [NILM wiki](http://wiki.nilm.eu/)? Thoughts @JackKelly @nipunbatra ?

**Comment by nipunbatra** [comment]
On the topic of specifying the number of states in the experiment, the following blog post should be helpful- http://nipunbatra.github.io/2015/07/nilmtk-states/

I fully agree that it is a good practice to give details about the number of states used for modelling.

I think we may need something like a dummy's guide to reproducible analysis in nilmtk? That should mention about the number of states used etc.

**Comment by yetanotherion** [comment]
@nipunbatra @oliparson would you like me to start writing a section about
- where does the exact_fhmm implementation fit in the literature, 
- why 
 - the number of states of each devices and 
 - the number of devices, 

is worth considering when benchmarking the performance of that implementation ?
